src/agent/Core/Controller/CheckoutSession.cpp

```diff
@@ -178,36 +179,110 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
+		Request *req = static_cast<Request *>(io->data);
+		Client *client = static_cast<Client *>(req->client);
+		Controller *that = static_cast<Controller *>(Controller::getServerFromClient(client));
+        ev_io_stop(that->getLoop(), io);
+
+		if (revents & EV_WRITE) {
+			// connected
+			try {
+				int connectError = 0;
+				socklen_t connectErrorLen = sizeof(connectError);
+				if (-1 == getsockopt(req->session->fd(), SOL_SOCKET, SO_ERROR, &connectError, &connectErrorLen)) {
+					int err = errno;
+					throw SystemException("Cannot check socket status", err);
+				} else if (connectError != 0) {
+					// connect_error uses the same error codes as errno
+					throw SystemException("Cannot connect socket", connectError);
+				}
+			} catch (const SystemException &e) {
+				handleInitiateError(e, req, client, that);
+				return;
+			}
+
+			that->finishInitiatingSession(client, req);
+        } else {
+			// something went very wrong
+			int err = errno;
+			SystemException e("Cannot connect socket", err);
+			handleInitiateError(e, req, client, that);
+			return;
+        }
+
+		that->unrefRequest(req, __FILE__, __LINE__);
+}
+
+void
+Controller::handleInitiateError(const SystemException &e, Request *req, Client* client, Controller* that) {
+	//TODO: only retry the session initiation if error is recoverable
+	if (req->sessionCheckoutTry < MAX_SESSION_CHECKOUT_TRY) {
+		SKC_DEBUG_FROM_STATIC(that, client, "Error initiating session (" << e.what() << "); retrying (attempt " << int(req->sessionCheckoutTry) << ")");
+		that->getContext()->libev->runLater(boost::bind(checkoutSessionLater, req));
+		// checkoutSessionLater unrefs req for us
+	} else {
+		string message = "error initiating a session (";
+		message.append(e.what());
+		message.append(")");
+		that->disconnectWithError(&client, message);
+		that->unrefRequest(req, __FILE__, __LINE__);
+	}
+}
+
+void
+Controller::finishInitiatingSession(Client *client, Request *req) {
+	TRACE_POINT();
+	SKC_DEBUG(client, "Session initiated: fd=" << req->session->fd());
+	req->appSink.reinitialize(req->session->fd());
+	req->appSource.reinitialize(req->session->fd());
+	/***************/
+	/***************/
+	reinitializeAppResponse(client, req);
+	sendHeaderToApp(client, req);
+}
+
 void
 Controller::initiateSession(Client *client, Request *req) {
 	TRACE_POINT();
 	req->sessionCheckoutTry++;
 	try {
-		req->session->initiate(false);
+		if (!req->session->initiate()) {
+			SKC_DEBUG(client, "Waiting on connection to finish, to initiate session appRoot=" << req->options.appRoot);
+
+			req->connectedWatcherTimout.data = req;
+			// TODO: get timeout from config var
+			ev_timer_init(&req->connectedWatcherTimout, onTimeout, 30, 0);
+			req->connectedWatcher.data = req;
+			ev_io_init(&req->connectedWatcher, onWritable, req->session->fd(), EV_WRITE);
+			// only one of timeout or ready handler should fire, so only ref once
+			refRequest(req, __FILE__, __LINE__);
```

Comment:
refRequest is only necessary when you're making a non-libev async operation that can't be cancelled, in order to keep req alive until the completion callback. Since libev async operations can be cancelled, use of refRequest is not necessary.